### PR TITLE
New: Nested Tabs - Tab icons [ED-7329]

### PR DESCRIPTION
**Pull request by TzviRabinovitch**

### CODE REVIEW

File: modules/tabs-v2/assets/scss/frontend.scss [hunk 1]
```diff
@@ -1,6 +1,21 @@
 //
```
Comment (iNewLegend) [inline modules/tabs-v2/assets/scss/frontend.scss:1]: Reformat according to the style.

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 2]
```diff
@@ -84,6 +91,46 @@ protected function register_controls() {
...
+				'skin' => 'inline',
+				'label_block' => false,
+				'condition' => [
+					'selected_icon[value]!' => '',
+				],
+			]
+		);
+
+		$repeater->add_control(
+			'_element_id',
```
Comment (iNewLegend) [inline modules/tabs-v2/widgets/tabs-v2.php:120]:
```suggestion
			'_element_id',
```
Why the underscore?

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 3]
```diff
@@ -371,10 +419,24 @@ protected function render() {
...
+			$icon_html = Icons_Manager::try_get_icon_html( $item['selected_icon'], [ 'aria-hidden' => 'true' ] );
+			$icon_active_html = $icon_html;
+			if ( $this->is_active_icon_exist( $item ) ) {
+				$icon_active_html = Icons_Manager::try_get_icon_html( $item['active_icon'], [ 'aria-hidden' => 'true' ] );
+			}
+			$tabs_title_html .= "<div $title_render_attributes>
+<span $tab_icon_attributes> $icon_html</span>
+<span $tab_icon_active_attributes> $icon_active_html</span>
+<span>$tab_title</span>
+</div>";
```
Comment (iNewLegend) [inline modules/tabs-v2/widgets/tabs-v2.php:439]: Fix formating

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 4]
```diff
@@ -429,4 +500,12 @@ protected function content_template() {
...
 		<?php
 	}
+
+	/**
+	 * @param $item
+	 * @return bool
+	 */
+	protected function is_active_icon_exist( $item ) {
+		return array_key_exists( 'active_icon', $item ) && ! empty( $item['active_icon'] ) && ! empty( $item['active_icon']['value'] );
+	}
```
Comment (iNewLegend) [inline modules/tabs-v2/widgets/tabs-v2.php:510]: Why it's protected?

**Review by iNewLegend — CHANGES_REQUESTED** [review]

File: modules/tabs-v2/assets/scss/frontend.scss [hunk 5]
```diff
@@ -64,8 +65,45 @@
...
+						top: 0;
+						transform: translateY(-100%);
+					  }
 
+					  &:after {
+						top: 100%;
+					  }
+					}
+				  }
+				}
```
Comment (iNewLegend) [inline modules/tabs-v2/assets/scss/frontend.scss:93]: Please fix bracket formating.

File: modules/tabs-v2/assets/scss/frontend.scss [hunk 6]
```diff
@@ -204,6 +242,29 @@
 		border-width: var(--tabs-v2-border-width)
 	}
 }
+.elementor-tab-title {
```
Comment (iNewLegend) [inline modules/tabs-v2/assets/scss/frontend.scss:245]: Please fix bracket formating.

File: modules/tabs-v2/widgets/tabs-v2.php (hunk 3 above)
Comment (iNewLegend) [reply modules/tabs-v2/widgets/tabs-v2.php:442]: Here is example: https://github.com/elementor/elementor/blob/master/includes/widgets/accordion.php#L561

File: includes/managers/icons.php [hunk 7]
```diff
@@ -42,6 +42,39 @@ private static function get_needs_upgrade_option() {
...
+		return self::get_icon_html( $icon, $attributes, $tag );
+	}
+
+	/**
+	 * @param array $icon
+	 * @param array $attributes
+	 * @param $tag
+	 * @return bool|mixed|string
+	 */
+	private static function get_icon_html( array $icon, array $attributes, $tag ) {
```
Comment (matipojo) [inline includes/managers/icons.php:65]: A static method should not be private

File: includes/managers/icons.php [hunk 8]
```diff
@@ -42,6 +42,39 @@ private static function get_needs_upgrade_option() {
...
+	 * @param $attributes
+	 * @param $tag
+	 * @return bool|mixed|string
+	 */
+	public static function try_get_icon_html( $icon, $attributes = [], $tag = 'i' ) {
+		if ( empty( $icon['library'] ) ) {
+			return '';
+		}
+
+		return self::get_icon_html( $icon, $attributes, $tag );
```
Comment (matipojo) [inline includes/managers/icons.php:56]: Always prefer to call current class method by using the `static` instead of `self`

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 9]
```diff
@@ -84,6 +91,46 @@ protected function register_controls() {
...
+				'skin' => 'inline',
+				'label_block' => false,
+				'condition' => [
+					'selected_icon[value]!' => '',
+				],
+			]
+		);
+
+		$repeater->add_control(
+			'_element_id',
```
Comment (matipojo) [inline modules/tabs-v2/widgets/tabs-v2.php:120]:
```suggestion
			'element_id',
```

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 10]
```diff
@@ -369,12 +419,28 @@ protected function render() {
...
+
+			$icon_html = Icons_Manager::try_get_icon_html( $item['selected_icon'], [ 'aria-hidden' => 'true' ] );
+			$icon_active_html = $icon_html;
+			if ( $this->is_active_icon_exist( $item ) ) {
+				$icon_active_html = Icons_Manager::try_get_icon_html( $item['active_icon'], [ 'aria-hidden' => 'true' ] );
+			}
 
-			$tabs_title_html .= "<div $title_render_attributes>$tab_title</div>";
+			$tabs_title_html .= "<div $title_render_attributes>";
+			$tabs_title_html .= "	<span $tab_icon_attributes> $icon_html</span>";
```
Comment (matipojo) [inline modules/tabs-v2/widgets/tabs-v2.php:440]: Prefer explicit character, for example: `"\t"`

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 11]
```diff
@@ -402,12 +468,21 @@ protected function content_template() {
 			var elementUid = view.getIDInt().toString().substr( 0, 3 ); #>
 			<div class="elementor-tabs-wrapper" role="tablist">
 				<# _.each( settings['tabs'], function( item, index ) {
-				var tabCount = index + 1,
-				tabUid = elementUid + tabCount,
+				var tabCount = index + 1;
+				tabUid = elementUid + tabCount;
```
Comment (matipojo) [inline modules/tabs-v2/widgets/tabs-v2.php:472]: Why?

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 12]
```diff
@@ -402,12 +468,21 @@ protected function content_template() {
...
 				tabTitleKey = 'tab-title-' + tabUid;
+				iconSelected = elementor.helpers.renderIcon( view, item.selected_icon, { 'aria-hidden': true }, 'i' , 'object' );
+				iconActive = iconSelected;
+				if ( '' !== item.active_icon.value ){
+					iconActive = elementor.helpers.renderIcon( view, item.active_icon, { 'aria-hidden': true }, 'i' , 'object' );
+				}
+				tabId= 'elementor-tab-title-' + tabUid;
+				if ( '' !== item._element_id ){
```
Comment (matipojo) [inline modules/tabs-v2/widgets/tabs-v2.php:480]: JSCS

File: modules/tabs-v2/assets/scss/frontend.scss [hunk 13]
```diff
@@ -154,6 +190,30 @@
...
+			.elementor-icon-active {
+				display: none;
+		  	}
+		}
+		.elementor-active {
+			.elementor-icon-active {
+				display: unset;
+			}
+			.elementor-icon-selected {
+		  		display: none;
```
Comment (matipojo) [inline modules/tabs-v2/assets/scss/frontend.scss:204]: Why hide them?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 14]
```diff
@@ -47,6 +48,12 @@ protected function get_default_children_elements() {
 					'_title' => __( 'Tab #2', 'elementor' ),
 				],
 			],
+			[
+				'elType' => 'container',
+				'settings' => [
+					'_title' => __( 'Tab #3', 'elementor' ),
+				],
+			],
```
Comment (iNewLegend) [inline modules/tabs-v2/widgets/tabs-v2.php:56]: I think it should not be part of the PR.

Comment (TzviRabinovitch) [reply modules/tabs-v2/widgets/tabs-v2.php:56]: @iNewLegend , yes , this is something you missed , so I added it 
I don't think its a bug deal , and it definitely should not block the PR

File: modules/tabs-v2/assets/scss/frontend.scss (hunk 13 above)
Comment (TzviRabinovitch) [reply modules/tabs-v2/assets/scss/frontend.scss:204]: when its not active (".elementor-tab-title .elementor-icon-active") => the  active icon is hidden 
when its active (".elementor-active") => the  active icon is unset and the selected ( the original ) is hidden 
@matipojo

File: modules/tabs-v2/widgets/tabs-v2.php (hunk 9 above)
Comment (TzviRabinovitch) [reply modules/tabs-v2/widgets/tabs-v2.php:120]: @matipojo, this is standard for CSS ID (in container , column & section )

File: modules/tabs-v2/widgets/tabs-v2.php (hunk 11 above)
Comment (TzviRabinovitch) [reply modules/tabs-v2/widgets/tabs-v2.php:472]: @matipojo , it did not work w/o it , i will recheck

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 15]
```diff
@@ -402,12 +468,21 @@ protected function content_template() {
...
 			<div class="elementor-tabs-wrapper" role="tablist">
 				<# _.each( settings['tabs'], function( item, index ) {
-				var tabCount = index + 1,
-				tabUid = elementUid + tabCount,
+				var tabCount = index + 1;
+				tabUid = elementUid + tabCount;
 				tabTitleKey = 'tab-title-' + tabUid;
+				iconSelected = elementor.helpers.renderIcon( view, item.selected_icon, { 'aria-hidden': true }, 'i' , 'object' );
+				iconActive = iconSelected;
+				if ( '' !== item.active_icon.value ){
```
Comment (matipojo) [inline modules/tabs-v2/widgets/tabs-v2.php:476]:
```suggestion
				if ( '' !== item.active_icon.value ){
```
```suggestion
				if ( '' !== item.active_icon.value ) {
```

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 16]
```diff
@@ -402,12 +468,21 @@ protected function content_template() {
...
+					iconActive = elementor.helpers.renderIcon( view, item.active_icon, { 'aria-hidden': true }, 'i' , 'object' );
+				}
+				tabId= 'elementor-tab-title-' + tabUid;
```
Comment (matipojo) [inline modules/tabs-v2/widgets/tabs-v2.php:479]:
```suggestion
				tabId = 'elementor-tab-title-' + tabUid;
```

File: modules/tabs-v2/widgets/tabs-v2.php (hunk 12 above)
Comment (matipojo) [inline modules/tabs-v2/widgets/tabs-v2.php:480]:
```suggestion
				if ( '' !== item._element_id ) {
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/tabs-v2/widgets/tabs-v2.php (hunk 10 above)
Comment (TzviRabinovitch) [reply modules/tabs-v2/widgets/tabs-v2.php:440]: Done

File: modules/tabs-v2/widgets/tabs-v2.php (hunk 9 above)
Comment (TzviRabinovitch) [reply modules/tabs-v2/widgets/tabs-v2.php:120]: Done

File: modules/tabs-v2/assets/scss/frontend.scss (hunk 13 above)
[the remark this answers is not in the document]
Comment (TzviRabinovitch) [reply modules/tabs-v2/assets/scss/frontend.scss:204]: Done

File: modules/tabs-v2/widgets/tabs-v2.php (hunk 14 above)
Comment (TzviRabinovitch) [reply modules/tabs-v2/widgets/tabs-v2.php:56]: Done

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 17]
```diff
@@ -84,6 +91,46 @@ protected function register_controls() {
 			],
 		] );
 
+		$repeater->add_control(
+			'selected_icon',
```
[the remark this answers is not in the document]
Comment (TzviRabinovitch) [reply modules/tabs-v2/widgets/tabs-v2.php:95]: Done

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 18]
```diff
@@ -84,6 +91,46 @@ protected function register_controls() {
...
+				'label' => esc_html__( 'Icon', 'elementor' ),
+				'type' => Controls_Manager::ICONS,
+				'fa4compatibility' => 'icon',
+				'skin' => 'inline',
+				'label_block' => false,
+			]
+		);
+
+		$repeater->add_control(
+			'active_icon',
```
[the remark this answers is not in the document]
Comment (TzviRabinovitch) [reply modules/tabs-v2/widgets/tabs-v2.php:106]: Done

File: modules/tabs-v2/assets/scss/frontend.scss [hunk 19]
```diff
@@ -64,8 +63,45 @@
...
+						transform: translateY(-100%);
+					  }
+
+					  &:after {
+						top: 100%;
+					  }
+					}
+				  }
+				}
+			  > .elementor-tab-desktop-title {
```
Comment (Nevoss) [inline modules/tabs-v2/assets/scss/frontend.scss:92]:
```suggestion
			  > .elementor-tab-title {
```

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: modules/tabs-v2/assets/scss/frontend.scss [hunk 20]
```diff
@@ -64,8 +63,45 @@
...
+				> .elementor-align-icon- {
```
[the remark this answers is not in the document]
Comment (TzviRabinovitch) [reply modules/tabs-v2/assets/scss/frontend.scss:93]: removed - its preparatory and not needed now

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 21]
```diff
@@ -402,12 +460,21 @@ protected function content_template() {
 			var elementUid = view.getIDInt().toString().substr( 0, 3 ); #>
 			<div class="elementor-tabs-wrapper" role="tablist">
 				<# _.each( settings['tabs'], function( item, index ) {
-				var tabCount = index + 1,
+				var tabCount = index + 1
```
Comment (Nevoss) [inline modules/tabs-v2/widgets/tabs-v2.php:463]: Why `,` was removed?

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 22]
```diff
@@ -369,12 +414,25 @@ protected function render() {
...
+			$tab_icon_active_attributes = $this->get_render_attribute_string( 'tab-icon-active' );
+
+			$icon_html = Icons_Manager::try_get_icon_html( $item['tab_icon'], [ 'aria-hidden' => 'true' ] );
+			$icon_active_html = $icon_html;
+			if ( $this->is_active_icon_exist( $item ) ) {
+				$icon_active_html = Icons_Manager::try_get_icon_html( $item['tab_icon_active'], [ 'aria-hidden' => 'true' ] );
+			}
 
-			$tabs_title_html .= "<div $title_render_attributes>$tab_title</div>";
+			$tabs_title_html .= "<div $title_render_attributes>";
```
Comment (Nevoss) [inline modules/tabs-v2/widgets/tabs-v2.php:431]:
```suggestion
			$tabs_title_html .= "<div {$title_render_attributes}>";
```

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 23]
```diff
@@ -369,12 +414,25 @@ protected function render() {
...
+			$tabs_title_html .= "\t<span $tab_icon_attributes> $icon_html</span>";
+			$tabs_title_html .= "\t<span $tab_icon_active_attributes> $icon_active_html</span>";
+			$tabs_title_html .= "\t<span>$tab_title</span>";
+			$tabs_title_html .= '</div>';
```
Comment (Nevoss) [inline modules/tabs-v2/widgets/tabs-v2.php:435]:
```suggestion
			$tabs_title_html .= "\t<span {$tab_icon_attributes}> $icon_html</span>";
			$tabs_title_html .= "\t<span {$tab_icon_active_attributes}>{$icon_active_html}</span>";
			$tabs_title_html .= "\t<span>{$tab_title}</span>";
			$tabs_title_html .= '</div>';
```

**Review by Nevoss — CHANGES_REQUESTED** [review]

**Review by Nevoss — CHANGES_REQUESTED** [review]

**Review by Nevoss — DISMISSED** [review]

File: modules/tabs-v2/assets/scss/frontend.scss [hunk 24]
```diff
@@ -154,6 +177,33 @@
...
+				display: unset;
+		  	}
+		}
+		.elementor-active {
+			.elementor-tab-icon {
+		  		display: none;
+			}
+		  .elementor-tab-icon-active {
+			display: unset;
+		  }
```
Comment (matipojo) [inline modules/tabs-v2/assets/scss/frontend.scss:195]: Fix indentation

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 25]
```diff
@@ -350,8 +393,10 @@ protected function render() {
 			$tab_title = $a11y_improvements_experiment ? $item['tab_title'] : '<a href="">' . $item['tab_title'] . '</a>';
 			$tab_title_mobile_setting_key = $this->get_repeater_setting_key( 'tab_title_mobile', 'tabs', $tab_count );
 
+			$tab_id = '' === $item['element_id'] ? 'elementor-tab-title-' . $id_int . $tab_count : $item['element_id'];
```
Comment (matipojo) [inline modules/tabs-v2/widgets/tabs-v2.php:396]: It should handle also if it's not set.
```suggestion
			$tab_id = empty( $item['element_id'] ) ? 'elementor-tab-title-' . $id_int . $tab_count : $item['element_id'];
```

File: modules/tabs-v2/widgets/tabs-v2.php [hunk 26]
```diff
@@ -369,12 +414,25 @@ protected function render() {
 				'tabindex' => 1 === $tab_count ? '0' : '-1',
 				'aria-controls' => 'elementor-tab-content-' . $id_int . $tab_count,
 				'aria-expanded' => 'false',
+				'id' => $tab_id,
```
Comment (matipojo) [inline modules/tabs-v2/widgets/tabs-v2.php:417]: It's a duplication of the `id` (desktop/mobile)

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/tabs-v2/widgets/tabs-v2.php (hunk 26 above)
Comment (TzviRabinovitch) [reply modules/tabs-v2/widgets/tabs-v2.php:417]: 2nd item changed to {element-id}-container

File: modules/tabs-v2/assets/scss/frontend.scss (hunk 24 above)
Comment (TzviRabinovitch) [reply modules/tabs-v2/assets/scss/frontend.scss]: Done
